### PR TITLE
Improve UX around setting Proxy configurations in the spec

### PR DESCRIPTION
**Pull request by maci3jka**
This PR introduces `httpProxy`, `httpsProxy`, `noProxy` parameters that adds support for proxy settings for contained in k8s. Provides functionality mentioned in [k8s docs](https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/proxy/).
KU-1801

### CODE REVIEW

File: bootstrap/api/v1beta2/ck8sconfig_types.go [hunk 1]
```diff
@@ -66,6 +66,18 @@ type CK8sConfigSpec struct {
...
+	// +optional
+	HTTPSProxy string `json:"httpsProxy,omitempty"`
+
+	// HTTPProxy is optional http proxy configuration
+	// +optional
+	HTTPProxy string `json:"httpProxy,omitempty"`
+
+	// NoProxy is optional no proxy configuration
+	// +optional
+	NoProxy string `json:"noProxy,omitempty"`
```
Comment (berkayoz) [inline bootstrap/api/v1beta2/ck8sconfig_types.go:79]: I'm not too happy with the naming here as these are only getting set for containerd. We should make it explicit like `ContainerdHTTPProxy`  if this is what we want.

**Review by berkayoz — COMMENTED** [review]
LGTM overall, I have one nit/question.

@bschimke95 Should we follow the k8s-operator approach of adding these variables to `/etc/environment` instead? If not we should make the variables reflect this is only for containerd.

File: pkg/cloudinit/controlplane_join_test.go [hunk 2]
```diff
@@ -82,6 +83,86 @@ func TestNewJoinControlPlane(t *testing.T) {
...
+			BootCommands:      []string{"bootcmd"},
+			PreRunCommands:    []string{"prerun1", "prerun2"},
+			PostRunCommands:   []string{"postrun1", "postrun2"},
+			ExtraFiles: []cloudinit.File{{
+				Path:        "/tmp/file",
+				Content:     "test file",
+				Permissions: "0400",
+				Owner:       "root:root",
+			}},
+			SnapstoreProxyScheme: "http",
```
Comment (eaudetcobello) [inline pkg/cloudinit/controlplane_join_test.go:101]: Let's remove snapstore proxy fields from the tests of containerd proxy, to make it clear that they are not related

File: pkg/cloudinit/controlplane_join_test.go [hunk 3]
```diff
@@ -82,6 +83,86 @@ func TestNewJoinControlPlane(t *testing.T) {
...
+		JoinToken: "test-token",
+	})
+
+	g.Expect(err).NotTo(HaveOccurred())
+
+	// Verify the boot commands.
+	g.Expect(config.BootCommands).To(Equal([]string{"bootcmd"}))
+
+	// Verify the run commands.
+	g.Expect(config.RunCommands).To(Equal([]string{
```
Comment (eaudetcobello) [inline pkg/cloudinit/controlplane_join_test.go:119]: It's not necessary to re-test all these items. Only test the relevant ones since we have a test for all of them already. 

Instead, only test the items that would be affected by setting a containerd proxy . For example, test that the containerd proxy setup script is present

**Review by eaudetcobello — CHANGES_REQUESTED** [review]
Looking good. Please apply my comments to the tests we added. We don't need to duplicate all the lines that are already tested by other tests, instead let's only check for the fields we know are relevant to containerd proxy configuration

File: pkg/cloudinit/controlplane_init_test.go [hunk 4]
```diff
@@ -105,6 +107,103 @@ func TestNewInitControlPlane(t *testing.T) {
 	), "Some /capi/scripts files are missing")
 }
 
+func TestNewInitControlPlaneWithOptionalProxySettings(t *testing.T) {
+	g := NewWithT(t)
+	for _, tc := range []struct {
```
Comment (maci3jka) [inline pkg/cloudinit/controlplane_init_test.go:112]: Is adding multiple test cases ok or too excessive?

Comment (berkayoz) [reply pkg/cloudinit/controlplane_init_test.go:112]: Checking it in 2 cases should be enough. One to see the files are appended when proxy is configured, and maybe one to see the files are not created when proxy is not configured.

File: pkg/cloudinit/common.go [hunk 5]
```diff
@@ -190,3 +201,32 @@ func getSnapstoreProxyConfigFiles(data BaseUserData) []File {
...
 }
+
+// getProxyConfigFiles returns the node proxy config files.
+// If the HTTPProxy or HTTPPSProxy is not set, it returns nil.
+// Nil indicates that no files are returned.
+func getProxyConfigFiles(data BaseUserData) []File {
+	if data.HTTPProxy == "" && data.HTTPSProxy == "" {
+		return nil
+	}
+	return []File{
```
Comment (berkayoz) [inline pkg/cloudinit/common.go:212]: Let's check if the variable is provided to add the respective file to the list. For example
```
if data.HTTPProxy != "" {
   files = append(files, File{...})
}
```

File: bootstrap/config/crd/bases/bootstrap.cluster.x-k8s.io_ck8sconfigs.yaml [hunk 6]
```diff
@@ -87,6 +84,9 @@ spec:
                     - secret
                     type: object
                 type: object
+              channel:
```
Comment (berkayoz) [inline bootstrap/config/crd/bases/bootstrap.cluster.x-k8s.io_ck8sconfigs.yaml:87]: Why does this change here?
Did we forget running `make generate && make manifests` before?

**Review by berkayoz — COMMENTED** [review]
Thank you, we're almost there some final comments

File: bootstrap/config/crd/bases/bootstrap.cluster.x-k8s.io_ck8sconfigs.yaml (hunk 6 above)
Comment (maci3jka) [reply bootstrap/config/crd/bases/bootstrap.cluster.x-k8s.io_ck8sconfigs.yaml:87]: looks like it `make generate && make manifests` created it like that for me

File: pkg/cloudinit/controlplane_init_test.go (hunk 4 above)
Comment (maci3jka) [reply pkg/cloudinit/controlplane_init_test.go:112]: I left 3 additonaly with one field to make sure that logic behaves as it sould

**Review by berkayoz — APPROVED** [review]
LGTM overall, one small nit before merging.

**Review by bschimke95 — APPROVED** [review]
LGTM

File: pkg/cloudinit/scripts/configure-proxy.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  echo "https_proxy=${HTTPS_PROXY}" >> "${ENVIRONMENT_FILE}"
+  echo "HTTPS_PROXY=${HTTPS_PROXY}" >> "${ENVIRONMENT_FILE}"
+fi
+
+
+if [ -f ${NO_PROXY_FILE} ]; then
+  NO_PROXY=$(cat ${NO_PROXY_FILE})
+  echo "no_proxy=${NO_PROXY}" >> "${ENVIRONMENT_FILE}"
+  echo "NO_PROXY=${NO_PROXY}" >> "${ENVIRONMENT_FILE}"
+fi
```
Comment (eaudetcobello) [inline pkg/cloudinit/scripts/configure-proxy.sh:31]: Can you make it so there are only 1 line between the if's? Same thing between lines 9 and 13.

**Review by eaudetcobello — APPROVED** [review]
